File: scripts/gen-icons.py
#!/usr/bin/env python3
"""
gen-icons.py — Generate placeholder app icons for Barq Cowork.

Usage:
    python3 scripts/gen-icons.py

Requirements:
    pip install Pillow

The script generates a lightning bolt icon in Barq's brand palette and
writes all required sizes to apps/desktop/src-tauri/icons/.
Replace these placeholders with professionally designed artwork before
publishing a release.
"""

import os
import sys
import struct
import zlib
import math

ICONS_DIR = os.path.join(
    os.path.dirname(__file__), "..", "apps", "desktop", "src-tauri", "icons"
)

# Brand colours
BG       = (18,  18,  27,  255)   # near-black
ACCENT   = (249, 115,  22, 255)   # orange-500
GLOW     = (251, 146,  60, 180)   # orange-400 semi-transparent
WHITE    = (255, 255, 255, 255)
TRANSP   = (0,   0,   0,   0)


def _write_chunk(chunk_type: bytes, data: bytes) -> bytes:
    chunk = chunk_type + data
    crc   = zlib.crc32(chunk) & 0xFFFFFFFF
    return struct.pack(">I", len(data)) + chunk + struct.pack(">I", crc)


def create_rgba_png(width: int, height: int, filepath: str) -> None:
    """Render a lightning bolt icon and save as RGBA PNG."""
    sig  = b"\x89PNG\r\n\x1a\n"
    ihdr = _write_chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0))

    cx, cy = width / 2, height / 2
    # Bolt vertices (normalised 0-1, then scaled)
    bolt = [
        (0.50, 0.05), (0.28, 0.52), (0.48, 0.52),
        (0.35, 0.95), (0.72, 0.42), (0.52, 0.42),
    ]
    bolt_px = [(x * width, y * height) for x, y in bolt]

    def point_in_polygon(px, py, polygon):
        n     = len(polygon)
        inside = False
        j     = n - 1
        for i in range(n):
            xi, yi = polygon[i]
            xj, yj = polygon[j]
            if ((yi > py) != (yj > py)) and (px < (xj - xi) * (py - yi) / (yj - yi) + xi):
                inside = not inside
            j = i
        return inside

    rows = []
    for y in range(height):
        row = bytearray(b"\x00")
        for x in range(width):
            dx, dy  = x - cx, y - cy
            dist    = math.hypot(dx, dy)
            r_outer = min(width, height) * 0.46
            r_inner = min(width, height) * 0.42

            if dist > r_outer:
                row += bytearray(TRANSP)
            elif dist > r_inner:
                # Circular border
                t    = (dist - r_inner) / (r_outer - r_inner)
                a    = int(255 * (1 - t))
                row += bytearray((ACCENT[0], ACCENT[1], ACCENT[2], a))
            elif point_in_polygon(x, y, bolt_px):
                row += bytearray(ACCENT)
            else:
                row += bytearray(BG)
        rows.append(bytes(row))

    compressed = zlib.compress(b"".join(rows), 9)
    idat = _write_chunk(b"IDAT", compressed)
    iend = _write_chunk(b"IEND", b"")

    os.makedirs(os.path.dirname(filepath), exist_ok=True)
    with open(filepath, "wb") as fh:
        fh.write(sig + ihdr + idat + iend)
    print(f"  ✓ {os.path.basename(filepath)}  ({width}×{height})")


def create_ico(png_32_path: str, ico_path: str) -> None:
    """Wrap the 32×32 PNG inside a minimal .ico container."""
    with open(png_32_path, "rb") as fh:
        png_data = fh.read()

    # ICO header (reserved=0, type=1, count=1)
    header  = struct.pack("<HHH", 0, 1, 1)
    # Directory entry for a 32×32 PNG image
    img_size   = len(png_data)
    img_offset = 6 + 16          # header + one directory entry
    entry = struct.pack("<BBBBHHII", 32, 32, 0, 0, 1, 32, img_size, img_offset)
    with open(ico_path, "wb") as fh:
        fh.write(header + entry + png_data)
    print(f"  ✓ {os.path.basename(ico_path)}")


def create_icns(png_128_path: str, icns_path: str) -> None:
    """Wrap a 128×128 PNG inside a minimal .icns container (ic07 type)."""
    with open(png_128_path, "rb") as fh:
        png_data = fh.read()

    # Each ICNS chunk: 4-byte OSType + 4-byte (chunk length including header) + data
    icns_type   = b"ic07"           # 128×128 PNG
    chunk_len   = 8 + len(png_data)
    chunk       = icns_type + struct.pack(">I", chunk_len) + png_data
    file_len    = 8 + len(chunk)    # 'icns' magic + total size + chunks
    header      = b"icns" + struct.pack(">I", file_len)
    with open(icns_path, "wb") as fh:
        fh.write(header + chunk)
    print(f"  ✓ {os.path.basename(icns_path)}")


def main() -> None:
    d = os.path.abspath(ICONS_DIR)
    print(f"Generating icons → {d}")

    p32  = os.path.join(d, "32x32.png")
    p128 = os.path.join(d, "128x128.png")
    p256 = os.path.join(d, "128x128@2x.png")
    ico  = os.path.join(d, "icon.ico")
    icns = os.path.join(d, "icon.icns")

    create_rgba_png(32,  32,  p32)
    create_rgba_png(128, 128, p128)
    create_rgba_png(256, 256, p256)
    create_ico(p32, ico)
    create_icns(p128, icns)

    print("\nDone. Replace these placeholders with professional artwork before release.")


if __name__ == "__main__":
    main()

File: scripts/gen-icons.go
package main

import (
	"bytes"
	"compress/zlib"
	"encoding/binary"
	"fmt"
	"hash/crc32"
	"math"
	"os"
	"path/filepath"
)

var (
	iconsDir = filepath.Join("apps", "desktop", "src-tauri", "icons")
	bg       = [4]byte{18, 18, 27, 255}
	accent   = [4]byte{249, 115, 22, 255}
	transp   = [4]byte{0, 0, 0, 0}
)

func main() {
	root, err := os.Getwd()
	if err != nil {
		panic(err)
	}
	dir := filepath.Join(root, iconsDir)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		panic(err)
	}

	fmt.Printf("Generating icons -> %s\n", dir)

	p32 := filepath.Join(dir, "32x32.png")
	p128 := filepath.Join(dir, "128x128.png")
	p256 := filepath.Join(dir, "128x128@2x.png")
	ico := filepath.Join(dir, "icon.ico")
	icns := filepath.Join(dir, "icon.icns")

	must(createRGBAIcon(32, 32, p32))
	must(createRGBAIcon(128, 128, p128))
	must(createRGBAIcon(256, 256, p256))
	must(createICO(p32, ico))
	must(createICNS(p128, icns))

	fmt.Println("Done. Replace these placeholders with production artwork before release.")
}

func must(err error) {
	if err != nil {
		panic(err)
	}
}

func createRGBAIcon(width, height int, path string) error {
	var rows bytes.Buffer
	cx, cy := float64(width)/2, float64(height)/2
	rOuter := math.Min(float64(width), float64(height)) * 0.46
	rInner := math.Min(float64(width), float64(height)) * 0.42
	bolt := [][2]float64{
		{0.50, 0.05}, {0.28, 0.52}, {0.48, 0.52},
		{0.35, 0.95}, {0.72, 0.42}, {0.52, 0.42},
	}

	for y := 0; y < height; y++ {
		rows.WriteByte(0)
		for x := 0; x < width; x++ {
			dist := math.Hypot(float64(x)-cx, float64(y)-cy)
			switch {
			case dist > rOuter:
				rows.Write(transp[:])
			case dist > rInner:
				alpha := uint8(255 * (1 - (dist-rInner)/(rOuter-rInner)))
				rows.Write([]byte{accent[0], accent[1], accent[2], alpha})
			case pointInPolygon(float64(x), float64(y), bolt, width, height):
				rows.Write(accent[:])
			default:
				rows.Write(bg[:])
			}
		}
	}

	var compressed bytes.Buffer
	zw := zlib.NewWriter(&compressed)
	if _, err := zw.Write(rows.Bytes()); err != nil {
		return err
	}
	if err := zw.Close(); err != nil {
		return err
	}

	var png bytes.Buffer
	png.Write([]byte{0x89, 'P', 'N', 'G', '\r', '\n', 0x1a, '\n'})
	png.Write(writePNGChunk("IHDR", mustBytes(func() ([]byte, error) {
		var b bytes.Buffer
		for _, v := range []uint32{uint32(width), uint32(height)} {
			if err := binary.Write(&b, binary.BigEndian, v); err != nil {
				return nil, err
			}
		}
		b.Write([]byte{8, 6, 0, 0, 0})
		return b.Bytes(), nil
	})))
	png.Write(writePNGChunk("IDAT", compressed.Bytes()))
	png.Write(writePNGChunk("IEND", nil))

	if err := os.WriteFile(path, png.Bytes(), 0o644); err != nil {
		return err
	}
	fmt.Printf("  created %s (%dx%d)\n", filepath.Base(path), width, height)
	return nil
}

func createICO(pngPath, icoPath string) error {
	data, err := os.ReadFile(pngPath)
	if err != nil {
		return err
	}
	var out bytes.Buffer
	if err := binary.Write(&out, binary.LittleEndian, uint16(0)); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, uint16(1)); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, uint16(1)); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, []byte{32, 32, 0, 0}); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, uint16(1)); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, uint16(32)); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, uint32(len(data))); err != nil {
		return err
	}
	if err := binary.Write(&out, binary.LittleEndian, uint32(22)); err != nil {
		return err
	}
	out.Write(data)
	return os.WriteFile(icoPath, out.Bytes(), 0o644)
}

func createICNS(pngPath, icnsPath string) error {
	data, err := os.ReadFile(pngPath)
	if err != nil {
		return err
	}
	chunkLen := uint32(8 + len(data))
	fileLen := uint32(8) + chunkLen

	var out bytes.Buffer
	out.WriteString("icns")
	if err := binary.Write(&out, binary.BigEndian, fileLen); err != nil {
		return err
	}
	out.WriteString("ic07")
	if err := binary.Write(&out, binary.BigEndian, chunkLen); err != nil {
		return err
	}
	out.Write(data)
	return os.WriteFile(icnsPath, out.Bytes(), 0o644)
}

func pointInPolygon(px, py float64, polygon [][2]float64, width, height int) bool {
	inside := false
	j := len(polygon) - 1
	for i := 0; i < len(polygon); i++ {
		xi, yi := polygon[i][0]*float64(width), polygon[i][1]*float64(height)
		xj, yj := polygon[j][0]*float64(width), polygon[j][1]*float64(height)
		if (yi > py) != (yj > py) && px < (xj-xi)*(py-yi)/(yj-yi)+xi {
			inside = !inside
		}
		j = i
	}
	return inside
}

func writePNGChunk(name string, data []byte) []byte {
	var out bytes.Buffer
	_ = binary.Write(&out, binary.BigEndian, uint32(len(data)))
	out.WriteString(name)
	out.Write(data)
	crc := crc32.ChecksumIEEE(append([]byte(name), data...))
	_ = binary.Write(&out, binary.BigEndian, crc)
	return out.Bytes()
}

func mustBytes(fn func() ([]byte, error)) []byte {
	b, err := fn()
	if err != nil {
		panic(err)
	}
	return b
}
